### PR TITLE
Linking to external projects

### PR DESCRIPTION
**Pull request by haraldkl**
This adds the possibility to resolve references to external FORD projects. For this to work the project, which is referred to needs to publish its entities in a json database, which is achieved by the `externalize: true` setting in the documentation configuration. Projects that want to make use of that and contain references to the project have to specify where to find them via the `external: ` setting in the configuration. Multiple external projects can be referred to.

An example for this is shown in the treelm project https://osdn.net/projects/apes/scm/hg/treelm/blobs/tip/tem_mainpage.md with the resulting documentation available at https://geb.sts.nt.uni-siegen.de/doxy/treelm/ an example where it makes use of the subproject aotus can be seen in https://geb.sts.nt.uni-siegen.de/doxy/treelm/module/tem_abortcriteria_module.html (the `aot_*` routines have links to another ford documentation).

This allows larger projects to be split into smaller projects and individually generating the documentation for the smaller projects.

### CODE REVIEW

File: ford/graphs.py [hunk 1]
```diff
@@ -245,23 +254,24 @@ def __init__(self,obj,gd,hist={}):
...
-                    n.visible = getattr(var.proto[0],'visible',True)
-                    if self in n.comp_of:
-                        n.comp_of[self] += ', ' + var.name
-                    else:
-                        n.comp_of[self] = var.name
-                    if n in self.comp_types:
-                        self.comp_types[n] += ', ' + var.name
-                    else:
-                        self.comp_types[n] = var.name
+            if not hasattr(obj, 'extURL'):
```
Comment (ZedThree) [inline ford/graphs.py:257]: For this sort of thing, I often prefer to flip the conditional and return, then the remainder doesn't need to be indented so much -- also helps with diffs. This isn't a blocker, just some polish :)

File: ford/fortran_project.py [hunk 2]
```diff
@@ -28,8 +28,7 @@
 
 import os
 import toposort
-
-import ford.sourceform
```
Comment (ZedThree) [inline ford/fortran_project.py:32]: This is a little surprising! Does this module not use `sourceform` directly?

File: ford/sourceform.py [hunk 3]
```diff
@@ -2532,3 +2537,50 @@ def get_name(self,item):
             return name
 
 namelist = NameSelector()
+
+
+class ExtModule(FortranModule):
```
Comment (ZedThree) [inline ford/sourceform.py:2542]: Rename suggestion for all these classes: `Ext*` -> `External*`

File: ford/sourceform.py [hunk 4]
```diff
@@ -2532,3 +2537,50 @@ def get_name(self,item):
...
+
+    def __init__(self):
+        self.name = ''
+        self.uses = []
+        self.pub_procs = {}
+        self.pub_absints = {}
+        self.pub_types = {}
+        self.pub_vars = {}
+        self.extURL = ''
```
Comment (ZedThree) [inline ford/sourceform.py:2551]: Rename suggestion for this variable: `extURL` -> `external_URL`

File: ford/utils.py [hunk 5]
```diff
@@ -24,9 +24,17 @@
...
 import os.path
+import json
+import sys
+import ford.sourceform
+if (sys.version_info[0] > 2):
+    from urllib.request import urlopen
+    from urllib.parse import urljoin
+else:
+    from urllib import urlopen
+    from urlparse import urljoin
```
Comment (ZedThree) [inline ford/utils.py:37]: Don't worry about supporting Python 2, it's dead :)

```suggestion
from urllib.request import urlopen
from urllib.parse import urljoin
```

File: ford/utils.py [hunk 6]
```diff
@@ -318,3 +332,135 @@ def sub_macros(string):
...
+        extObj.name = extDict['name']
+        if extDict['extURL']:
+            extDict['extURL'] = '/' + extDict['extURL'].split('/', 1)[-1]
+            extObj.extURL = url + extDict['extURL']
+        else:
+            extObj.extURL = extDict['extURL']
+        extObj.obj = extDict['obj']
+        extObj.parent = parent
+        for key in attribs:
+            if key in extDict:
```
Comment (ZedThree) [inline ford/utils.py:424]: Here, I'd be tempted to do:
```suggestion
            if key not in extDict:
                continue
```
Then the rest of the loop can be indented less

**Review by ZedThree — APPROVED** [review]
Really nice, thanks @haraldkl ! Could be the start of being able to interface to other backends like Sphinx, which would be really cool.

I've made a few suggestions that you may wish to consider, but I'm happy to merge as-is

File: ford/fortran_project.py (hunk 2 above)
Comment (haraldkl) [reply ford/fortran_project.py:32]: I'm sorry might be an oversight from the merging. I am not sure, why it is removed.

File: ford/graphs.py (hunk 1 above)
Comment (haraldkl) [reply ford/graphs.py:257]: Sure. Please make any adjustments as you see fit.

File: ford/sourceform.py (hunk 3 above)
Comment (haraldkl) [reply ford/sourceform.py:2542]: Yes, fine with me.

File: ford/sourceform.py (hunk 4 above)
Comment (haraldkl) [reply ford/sourceform.py:2551]: Fine by me.

File: ford/utils.py (hunk 5 above)
Comment (haraldkl) [reply ford/utils.py:37]: Well, the change is several years old, and back then we still cared about Python 2 ;) It's fine to do away with it.

File: ford/utils.py (hunk 6 above)
Comment (haraldkl) [reply ford/utils.py:424]: That's also fine by me.

**Comment by haraldkl** [comment]
Do you want me to make those changes, or do you already have the diffs?

**Comment by ZedThree** [comment]
I've got the diffs, I'll make the changes now

**Comment by haraldkl** [comment]
Wonderful. Thanks a lot, you are working at lightning speed... ;)

**Comment by ZedThree** [comment]
@haraldkl I've pushed my changes to `external-changes` branch -- please could you give it a quick test to check I've not broken things?

Ideally we'd get some actual tests into the repo as well, but I think I need to come up with a good way of both creating test projects, and also verifying that they work. It's a bit late in the day for me to do that.

**Comment by haraldkl** [comment]
Looks like I am missing the markdown dependency now on the system we use. I'll need to update. Takes me a little while, sorry.

**Comment by haraldkl** [comment]
It fails for me, because the external url provided in `project.external` seems to yield only the first character.

Definition in the configuration: `external: aoturl = https://geb.sts.nt.uni-siegen.de/doxy/aotus`.

Traceback:

    Traceback (most recent call last):
      File "/usr/bin/ford", line 33, in <module>
        sys.exit(load_entry_point('FORD==6.0.0', 'console_scripts', 'ford')())
      File "/usr/lib/python3.9/site-packages/FORD-6.0.0-py3.9.egg/ford/__init__.py", line 450, in run
        main(proj_data,proj_docs,md)
      File "/usr/lib/python3.9/site-packages/FORD-6.0.0-py3.9.egg/ford/__init__.py", line 391, in main
       project.correlate()
      File "/usr/lib/python3.9/site-packages/FORD-6.0.0-py3.9.egg/ford/fortran_project.py", line 170, in correlate
       ford.utils.external(self)
      File "/usr/lib/python3.9/site-packages/FORD-6.0.0-py3.9.egg/ford/utils.py", line 448, in external
        url, short = register_macro(urldef)
     File "/usr/lib/python3.9/site-packages/FORD-6.0.0-py3.9.egg/ford/utils.py", line 308, in register_macro
       raise RuntimeError('Error, no alias name provided for {0}'.format(string))
   RuntimeError: Error, no alias name provided for a

I didn't dive into it yet to see where this comes from, maybe the handling of the config parsing has changed with the markdown update?

**Comment by haraldkl** [comment]
OK, the first problem was, due to a mistake in my merging: the external option got removed from the list options. The second problem arises from the change in the register_macro routine: this is used in the external routine and there the returned tuple is actually used. Another issue in the register_macro routine now was, that the value contained whitespaces, which prohibited the URL resolution due to a leading blank.

With the three added commits and the things from `external-changes` everything seems to work now.

**Comment by ZedThree** [comment]
Awesome, thanks @haraldkl ! I've managed to add a basic test that this works too, so we'll have some confidence that we won't accidentally break it in the future.

**Comment by haraldkl** [comment]
Great, thanks a lot!

**Pull request merged**